src/components/manage/index.js

```diff
@@ -163,29 +163,56 @@ export class Manage extends Component {
 
 
   canFinalize = () => {
-    const { contractStore, match } = this.props
+    const { contractStore } = this.props
 
-    return new Promise(resolve => {
+    return new Promise((resolve, reject) => {
       if ((!this.state.crowdsaleHasEnded) || (this.state.shouldDistribute && this.state.canDistribute)) {
         this.setState({ canFinalize: false })
         resolve(this.state.canFinalize)
 
       } else {
-        attachToContract(contractStore.crowdsale.abi, match.params.crowdsaleAddress)
-          .then(crowdsaleContract => { // eslint-disable-line no-loop-func
-            console.log('attach to crowdsale contract')
-
-            if (!crowdsaleContract) return Promise.reject('No contract available')
-            crowdsaleContract.methods.finalized().call((err, finalized) => {
-              return finalized;
-            }).then((finalized) => {
-              this.setState({ canFinalize: !finalized })
-              resolve(this.state.canFinalize)
-            })
-            .catch(() => {
-              this.setState({ canFinalize: false })
-              resolve(this.state.canFinalize)
-            })
+        const lastCrowdsaleAddress = contractStore.crowdsale.addr.slice(-1)[0]
+
+        attachToContract(contractStore.crowdsale.abi, lastCrowdsaleAddress)
+          .then(crowdsaleContract => {
+            const whenTokenContract = crowdsaleContract.methods.token().call()
+              .then(tokenAddress => attachToContract(contractStore.token.abi, tokenAddress))
+
+            return Promise.all([crowdsaleContract, whenTokenContract])
+              .then(([crowdsaleContract, tokenContract]) => {
+                if (!crowdsaleContract) {
+                  reject('No contract available')
+                } else {
+                  Promise.all([
+                    crowdsaleContract.methods.finalized().call(),
+                    crowdsaleContract.methods.isCrowdsaleFull().call(),
+                    crowdsaleContract.methods.areReservedTokensDistributed().call(),
+                    tokenContract.methods.reservedTokensDestinationsLen().call()
+                  ])
+                    .then(([finalized, isCrowdsaleFull, areTokensDistributed, reservedTokensLen]) => {
+                      let _canFinalize
+
+                      if (isCrowdsaleFull && !finalized) {
+                        // If all the tokens were sold and the crowdsale is not already finalized,
+                        // it can be finalized if there are no tokens to distribute, or if they were already distributed
+                        _canFinalize = parseInt(reservedTokensLen, 10) === 0 || areTokensDistributed
+                      } else {
+                        _canFinalize = false
```

Comment:
Let's imagine that crowdsale is not finalized, ended, but hard cap is not sold, reserved tokens are distributed. Why can we not to finalize?

Reply:
The `finalize()` method in the crowdsale contract has na `inState(State.Success)` modifier, and, if I'm understanding this correctly, for the crowdsale to be in `State.Success` the minimum goal has to have been reached. Am I missing something?

Reply:
Right, for crowdsale in `State.Success` the minimum goal is reached. But if the minimum goal is reached, it doesn't mean that crowdsale is in `State.success` state.

Reply:
Oh, ok. Should we just check that `getState() === 4` then?

Reply:
I think the full logic schematically should be rewritten as:
- if (isCrowdsaleFull && ((shouldDistribute && !canDistribute) || !shouldDistribute) => canFinalize = true
- else if (crowdsaleHasEnded && ((shouldDistribute && !canDistribute) || !shouldDistribute) => canFinalize =true
- else canFinalize = false

Reply:
@vbaranov what you think about it now? I'd tested it with different scenarios and worked fine.